src/psyclone/psyir/nodes/dynamic_omp_task_directive.py

```diff
@@ -0,0 +1,1977 @@
+        :type shared_list: List[:py:class:`psyclone.psyir.nodes.Reference`]
+        :param in_list: The input References for this task.
+        :type in_list: List[:py:class:`psyclone.psyir.nodes.Reference`]
+        :param out_list: The output References for this task.
+        :type out_list: List[:py:class:`psyclone.psyir.nodes.Reference`]
+        """
+        # For the node, check if it is Loop, Assignment or IfBlock
+        if isinstance(node, Assignment):
+            # Resolve assignment
+            self._evaluate_assignment(
+                node,
+                private_list,
+                firstprivate_list,
+                shared_list,
+                in_list,
+                out_list,
+            )
+        elif isinstance(node, Loop):
+            # Resolve loop
+            self._evaluate_loop(
+                node,
+                private_list,
+                firstprivate_list,
+                shared_list,
+                in_list,
+                out_list,
+            )
+        elif isinstance(node, IfBlock):
+            # Resolve IfBlock
+            self._evaluate_ifblock(
+                node,
+                private_list,
+                firstprivate_list,
+                shared_list,
+                in_list,
+                out_list,
+            )
+
+        # All other node types are ignored as they shouldn't affect
+        # dependency computation, as these are the only nodes that
+        # have read or write accesses that can get to this function
+        # as Calls are prohibited in validation.
+
+    def _compute_clauses(self):
+        """
+        Computes the clauses for this OMPTaskDirective.
+
+        The OMPTaskDirective must have exactly 1 child, which must be a Loop.
+        Upon confirming this, the function calls _evaluate_node to compute all
+        data-sharing attributes and dependencies.
+        The clauses are then built up from those, and returned.
+
+        :raises GenerationError: If the OMPTaskDirective has multiple children.
+        :raises GenerationError: If the OMPTaskDirective's child is not a Loop.
+
+        :returns: The clauses computed for this OMPTaskDirective.
+        :rtype: List[:py:class:`psyclone.psyir.nodes.OMPPrivateClause`,
+                     :py:class:`psyclone.psyir.nodes.OMPFirstprivateClause`,
+                     :py:class:`psyclone.psyir.nodes.OMPSharedClause`,
+                     :py:class:`psyclone.psyir.nodes.OMPDependClause`,
+                     :py:class:`psyclone.psyir.nodes.OMPDependClause`]
+        """
+
+        # These lists will store PSyclone nodes which are to be added to the
+        # clauses for this OMPTaskDirective.
+        private_list = []
+        firstprivate_list = []
+        shared_list = []
+        in_list = []
+        out_list = []
+
+        # Reset this in case we already computed clauses before but are
+        # recomputing them (usually due to a code change or multiple outputs).
```

Comment:
I'm not sure if it does - though perhaps this is where Sergi and I's previous discussions made an error.
Since the `_compute_clauses` call only happens at lowering, I don't think the tree should be modified after the clauses are computed.
The existence of `_update_node` functionality does mean that maybe that decision isn't necessary, however calling `compute_clauses` every time the tree is updated for these nodes may be expensive.

I think in theory the node that would need `update_node` (if the tree can change post-lowering) is the `OMPDirective` itself, however it has none of the functionality to compute any of the clauses, so it would have to replace itself with an instance of this class, then relower that node.

Do you think tree-changes post lowering can happen or are we ok?

Reply:
I don't think we can assume that we'll never modify the tree after lowering - there may be some use case down the line. Is it possible to implement an `update_node` that just raises a `NotImplementedError`? If that works then great. If on the other hand, it already gets triggered for reasons I haven't thought of then we may have to think of other ways of ensuring this doesn't happen.

Reply:
(Or, worst case, we document this as a limitation for now.)

Reply:
I'll add this to omp_task_directive and see how that goes. Dynamic ones shouldn't exist post lowering, and they don't need to do anything until lowering so I won't add it there

Reply:
Ok so this fails because upon initialising a node it calls _update_node on itself resulting in the error.
@arporter where is best to document this?